pyadjoint/checkpointing.py

```diff
@@ -255,37 +257,49 @@ def process_operation(self, cp_action, bar, **kwargs):
 
     @process_operation.register(Forward)
     def _(self, cp_action, bar, functional=None, **kwargs):
-        for step in cp_action:
+        step = cp_action.n0
+        while step in cp_action and step < self.total_timesteps:
```

Comment:
Why step < self.total_timesteps?

Reply:
This is because `SingleMemoryStorageSchedule` and `NoneCheckpointSchedule` have the forward schedule `Forward(0, sys.maxsize, False, True, StorageType.WORK)`.